server/src/lib.rs

```diff
@@ -1009,6 +1066,101 @@ mod tests {
+    // mock remote servers actually receive write entry events.
+    //
+    // This test is theoretically flaky, low probability though (in the order of 1e-30)
+    #[tokio::test]
+    async fn write_entry_downstream() {
+        const TEST_SHARD_ID: ShardId = 1;
+        const GOOD_REMOTE_ID_1: WriterId = 1;
+        const GOOD_REMOTE_ID_2: WriterId = 2;
+        const BAD_REMOTE_ID: WriterId = 666;
+        const GOOD_REMOTE_ADDR_1: &str = "http://localhost:111";
+        const GOOD_REMOTE_ADDR_2: &str = "http://localhost:222";
+        const BAD_REMOTE_ADDR: &str = "http://localhost:666";
+
+        let mut manager = TestConnectionManager::new();
+        let written_1 = Arc::new(AtomicBool::new(false));
+        manager.remotes.insert(
+            GOOD_REMOTE_ADDR_1.to_owned(),
+            Arc::new(TestRemoteServer {
+                written: Arc::clone(&written_1),
+            }),
+        );
+        let written_2 = Arc::new(AtomicBool::new(false));
+        manager.remotes.insert(
+            GOOD_REMOTE_ADDR_2.to_owned(),
+            Arc::new(TestRemoteServer {
+                written: Arc::clone(&written_2),
+            }),
+        );
+
+        let server = Server::new(manager, config());
+        server.set_id(NonZeroU32::new(1).unwrap()).unwrap();
+
+        let name = DatabaseName::new("foo".to_string()).unwrap();
+        server
+            .create_database(DatabaseRules::new(name), server.require_id().unwrap())
+            .await
+            .unwrap();
+
+        let remote_ids = vec![BAD_REMOTE_ID, GOOD_REMOTE_ID_1, GOOD_REMOTE_ID_2];
+        let db_name = DatabaseName::new("foo").unwrap();
+        let db = server.db(&db_name).unwrap();
+        {
+            let mut rules = db.rules.write();
+            rules.shard_config = Some(ShardConfig {
+                hash_ring: Some(HashRing {
+                    shards: vec![TEST_SHARD_ID].into(),
+                    ..Default::default()
+                }),
+                shards: Arc::new(
+                    vec![(TEST_SHARD_ID, remote_ids.clone())]
+                        .into_iter()
+                        .collect(),
+                ),
+                ..Default::default()
+            });
+        }
+
+        let line = "cpu bar=1 10";
+        let lines: Vec<_> = parse_lines(line).map(|l| l.unwrap()).collect();
+
+        let err = server.write_lines(&db_name, &lines).await.unwrap_err();
+        assert!(
+            matches!(err, Error::NoRemoteConfigured { node_group } if node_group == remote_ids)
+        );
+
+        // one remote is configured but it's down and we'll get connection error
+        server.update_remote(BAD_REMOTE_ID, BAD_REMOTE_ADDR.into());
+        let err = server.write_lines(&db_name, &lines).await.unwrap_err();
+        assert!(
+            matches!(err, Error::NoRemoteReachable { errors } if matches!(errors[BAD_REMOTE_ADDR], ConnectionManagerError::RemoteServerConnectError{..}))
```

Comment:
rustfmt doesn't touch text inside macro calls; I'd make this line a bit shorter and look a bit more like regular Rust code with something like:

```suggestion
            matches!(
                err, 
                Error::NoRemoteReachable { errors } if matches!(
                    errors[BAD_REMOTE_ADDR], 
                    ConnectionManagerError::RemoteServerConnectError { .. }
                )
            )
```

Reply:
thanks fixed